### ISSUE TITLE
Small score difference between target and "background"/ OOD class

### ISSUE THREAD

**Issue by mariana200196** [dlmacedo/distinction-maximization-loss#2]
Thank you very much for your research contribution. An easy-to-use loss function that can differentiate between target classes and "background"/ OOD classes is extremely useful. 

I want to use your solution so that my classifier, trained to predict the gender of cartoon faces, ignores images which don't contain a face (e.g., predicts the gender of a "background" or non-face image with a very low confidence). Unfortunately, I'm finding that my model doesn't greatly differentiate between faces and "background" images.

For example, the tensors are very similar in value:
![image](https://user-images.githubusercontent.com/91416518/172047089-fb937970-4628-41d2-9538-6abcd85482c0.png) ![image](https://user-images.githubusercontent.com/91416518/172047096-8fb547b5-c0af-45de-a75c-d45dec3a2e3d.png)

![image](https://user-images.githubusercontent.com/91416518/172047109-39b58b67-8eca-4de8-b7a6-d69cf24224ef.png) ![image](https://user-images.githubusercontent.com/91416518/172047121-f2c4c137-dfd9-4651-8e26-963e2892d2c4.png)

Where am I going wrong? I greatly appreciate any feedback you can give me.

**How I train the model**
Example batch:
![image](https://user-images.githubusercontent.com/91416518/172047331-2bb81dfa-13b6-4a42-9afb-f3c92cf6bd74.png)
```
def get_lr(optimizer):
    for param_group in optimizer.param_groups:
        return param_group['lr']

def train_model(model, criterion, optimizer, scheduler, num_epochs=25):
    since = time.time()

    best_model_wts = copy.deepcopy(model.state_dict())
    best_acc = 0.0
    history = []

    for epoch in range(num_epochs):
        result = {}
        epoch_start = time.time()
        print('Epoch {}/{}'.format(epoch+1, num_epochs))
        print('-' * 10)

        # each epoch has a training and validation phase
        for phase in ['train', 'val']:
            if phase == 'train':
                model.train()  # Set model to training mode
            else:
                model.eval()   # Set model to evaluate mode

            running_loss = 0.0
            running_corrects = 0
            correct_females = 0
            correct_males = 0

            # iterate over data
            for inputs, labels in dataloaders[phase]:
                inputs = inputs.to(device)
                labels = labels.to(device)

                # zero the parameter gradients
                optimizer.zero_grad()

                # forward
                # track history if only in train
                with torch.set_grad_enabled(phase == 'train'):
                    inputs, labels = criterion.preprocess(inputs, labels) #dismax

                    outputs = model(inputs)
                    _, preds = torch.max(outputs, 1)
                    loss = criterion(outputs, labels)

                    # backward + optimize only if in training phase
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()

                # statistics
                running_loss += loss.item() * inputs.size(0)
                running_corrects += torch.sum(preds == labels.data)
                for i, pred in enumerate(preds):
                    if labels[i] == 0 and labels[i] == pred: 
                      correct_females += 1
                    if labels[i] == 1 and labels[i] == pred:
                      correct_males += 1
            
            if phase == 'train':
                # record & update learning rate
                result['lrs'] = get_lr(optimizer)
                scheduler.step()

            # for printing to command line
            epoch_loss = running_loss / dataset_sizes[phase]
            epoch_acc = running_corrects.double().item() / dataset_sizes[phase]
            female_acc = correct_females / gender_dist[phase]['females']
            male_acc = correct_males / gender_dist[phase]['males']
            # for plotting
            result[phase+'_loss'] = epoch_loss
            result[phase+'_acc'] = epoch_acc
            result[phase+'_female_acc'] = female_acc
            result[phase+'_male_acc'] = male_acc

            # deep copy the model
            if phase == 'val' and epoch_acc > best_acc:
                best_acc = epoch_acc
                best_model_wts = copy.deepcopy(model.state_dict())
                # Save if the model has best accuracy till now
                torch.save(model, data_dir+'_model_best.pt')

            # print stats to command line
            print(f'{phase} Loss: {epoch_loss:.4f} Acc: {epoch_acc:.4f}')
            total_females = gender_dist[phase]['females']
            total_males = gender_dist[phase]['males']
            print(f'Female Acc: {correct_females}/{total_females} Male Acc: {correct_males}/{total_males}')
            print(f'Female Acc: {female_acc:.4f} Male Acc: {male_acc:.4f}')

            if phase == 'val':
                epoch_time = time.time() - epoch_start
                print('Time per epoch {:.0f}m {:.0f}s'.format(
                    epoch_time // 60, epoch_time % 60))

        history.append(result)
        print()

    time_elapsed = time.time() - since
    print('Training complete in {:.0f}m {:.0f}s'.format(
        time_elapsed // 60, time_elapsed % 60))
    print('Best val Acc: {:4f}'.format(best_acc))

    torch.save(history, data_dir+'_history.pt')

    # load best model weights
    model.load_state_dict(best_model_wts)
    return model, history

# Load pretrained ResNetXX Model
# dismax: https://github.com/dlmacedo/distinction-maximization-loss
import losses # dismax
model = models.resnet50(pretrained=True)

num_ftrs = model.fc.in_features
model.fc = losses.DisMaxLossFirstPart(num_ftrs, len(class_names)) # dismax
model = model.to(device)
criterion = losses.DisMaxLossSecondPart(model.fc) # dismax

# Observe that all parameters are being optimized
optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9) #lr=0.003

# Freeze until 2nd block (included): https://raminnabati.com/post/002_adv_pytorch_freezing_layers/
ct = 0
for child in model.children():
  ct += 1
  if ct < 6:
      for param in child.parameters():
          param.requires_grad = False

# Decay LR
exp_lr_scheduler = lr_scheduler.StepLR(optimizer, step_size=15, gamma=0.1)

# Train
model, history = train_model(model, criterion, optimizer, exp_lr_scheduler, num_epochs=30
```

**How I evaluate the model (one image at a time)**
```
## Make predictions, and save images with title
import torch.nn.functional as F
import pandas as pd
model.eval()

with torch.no_grad():
  for x in [f'{data_dir}']:
    # batch size 1 so can get file names correctly
    for i, (inputs, labels) in enumerate(dataloaders[x]):
      inputs = inputs.to(device)

      outputs = model(inputs)
      scores = model.fc.scores(outputs).item() # dismax 
      _, preds = torch.max(outputs, 1)
      path, _ = dataloaders[x].dataset.samples[i]
      fname = path.split('/')[-1]

      for j in range(inputs.size()[0]):
        ax = plt.subplot(1, 1, 1)
        ax.axis('off')

        pred = class_names[preds[j]]
        ax.set_title(f'Predicted Gender: {pred} ({scores}%)') # conf:.2f

        inp = inputs.cpu().data[j].numpy().transpose((1, 2, 0))
        mean = np.array([0.485, 0.456, 0.406])
        std = np.array([0.229, 0.224, 0.225])
        inp = std * inp + mean # denormalise image
        inp = np.clip(inp, 0, 1)
        ax.imshow(inp) # has to show otherwise canvas is blank
        plt.savefig(f'{DST_PATH}/{pred}/{fname}')
        plt.clf
```

**Comment by dlmacedo** [comment]
You need to have a threshold and compare the score with it.

Please, see this:

https://github.com/dlmacedo/entropic-out-of-distribution-detection/issues/1

Or simply use a precomputed threshold provided by the DisMax combined with the MMLES score.

Do not worry about the fact that the values of the tensors are close. This is normal as we are dealing with high entropy.

For example, looking at your results, it appears that a threshold above -6.46 and below -6.41 (for example, -6.43) would have worked fine to distinguish faces from no-faces in your case.

We suggest you may try also first IsoMax, IsoMax+, or DisMax dagger.
